### DISCUSSION TITLE
Using Podman Desktop on Linux with newer version in Podman Machine

### DISCUSSION THREAD

**Discussion by afbjorklund** [podman-desktop/podman-desktop#5762] [Ideas]
Currently, Podman Desktop doesn't allow creating a new machine on Linux. The feature is only available on Mac and Windows.

And the available version (Podman v3) is too old anyway, which means that not all the `podman machine` flags are compatible...

EDIT: New version of Podman Desktop doesn't require patching, but it doesn't allow creating machines (on Linux)

<details>
But you can patch the code, to allow using it anyway:

```diff
index fc1206e89..3c1b1300b 100644
--- a/extensions/podman/src/extension.ts
+++ b/extensions/podman/src/extension.ts
@@ -992,7 +992,7 @@ export async function activate(extensionContext: extensionApi.ExtensionContext):
   await registerUpdatesIfAny(provider, installedPodman, podmanInstall);
 
   // register autostart if enabled
-  if (isMac() || isWindows()) {
+  if (isMac() || isWindows() || isLinux()) {
     try {
       await updateMachines(provider);
     } catch (error) {
@@ -1035,7 +1035,7 @@ export async function activate(extensionContext: extensionApi.ExtensionContext):
   extensionContext.subscriptions.push(provider);
 
   // allows to create machines
-  if (isMac() || isWindows()) {
+  if (isMac() || isWindows() || isLinux()) {
     provider.setContainerProviderConnectionFactory({
       initialize: () => createMachine({}, undefined),
       create: createMachine,
```
</details>

And then you can download and install a newer version of `podman-remote`, installed as `podman-remote-static`:

https://github.com/containers/podman/releases

* `podman-remote-static-linux_amd64.tar.gz`
* `podman-remote-static-linux_arm64.tar.gz`

Installed into `/usr/local/bin/podman-remote-static`

```
podman-remote-static machine init
podman-remote-static machine start
```

And configure Podman Desktop, to use the newer `podman-remote-static` instead of the old `podman-remote`:

![podman-desktop-preferences-podman](https://github.com/containers/podman-desktop/assets/10364051/409b4daa-57de-4091-b5b1-32401138c6a4)

Now, Podman Desktop will show both the old Podman version as well as the new Podman Machine virtual machine:

![podman-desktop-machine-linux](https://github.com/containers/podman-desktop/assets/10364051/1ce4bff8-58d4-4afa-8014-90dd4427cb4c)

This means that all the new features will work, just as they do on macOS and Windows. And the old one is still there.

Note: Currently the user interface shows the **client** version for both connections, instead of the actual server version.

----

There will be some warnings about new features in the configuration, but those are mostly harmless (but annoying).

`WARN[0000] Failed to decode the keys ["engine.service_destinations.podman-machine-default.is_machine" "engine.service_destinations.podman-machine-default-root.is_machine". `

You will also need to set CONTAINER_CONNECTION to "podman-machine-default", when running podman-remote-static.

Otherwise it will try to connect to the system podman socket (like podman-remote), but fail to use it because of the version:

`Error: unable to connect to Podman socket: server API version is too old. Client "4.0.0" server "3.4.2"`

----

EDIT: Besides podman, you will also need to install `gvproxy` for networking: https://github.com/containers/gvisor-tap-vsock

https://github.com/containers/gvisor-tap-vsock/releases

* `gvproxy-linux-amd64`
* `gvproxy-linux-arm64`

Installed into `/usr/local/bin/gvproxy`

And also the new `virtiofsd` program:
https://gitlab.com/virtio-fs/virtiofsd

And make sure that you are `kvm-ok`, for running virtual machines: https://help.ubuntu.com/community/KVM/Installation

```console
$ sudo apt install -y cpu-checker qemu-kvm
$ kvm-ok
INFO: /dev/kvm exists
KVM acceleration can be used
```

On older distributions, you will also have to install a newer version of [QEMU](https://qemu.org)

**Comment by afbjorklund** [comment]
For some of the Linux distributions, the Podman version might be new enough* to not need a virtual machine at all...

Distro | Podman
--- | ---
Debian 11 | 3.0.1
Ubuntu 20.04 LTS | 3.4.2
Ubuntu 22.04 LTS | 3.4.4
Debian 12 | 4.3.1
AlmaLinux 8 (el8) | 4.6.1*
AlmaLinux 9 (el9) | 4.6.1*
Fedora 38 | 4.8.3*
Fedora 39 | 4.9.0*

https://repology.org/project/podman

**Reply by afbjorklund** [reply]
I was able to build Podman v4 for Debian (and Ubuntu LTS), using Podman v3 and the docker build scripts from nerdctl

https://github.com/afbjorklund/podman-full | mostly from https://podman.io/docs/installation#building-from-source

**Comment by cdrage** [comment]
Great idea and thank you for opening up the issue! This would also allow us to create a "rootful" / root podman machine on Linux more easily.

We'll work on adding this: https://github.com/containers/podman-desktop/issues/5823

**Reply by afbjorklund** [reply]
I did open a feature for it, but it didn't seem like something that the team was interested in adding

* https://github.com/containers/podman-desktop/issues/3068#issuecomment-1631008951
